### PR TITLE
Policy decay effect - WIP

### PR DESCRIPTION
**Pull request by AlexisOlson**
Decay the effect of policy with additional nodes.

Instead of P use:
image

Example P effect:
image

A coefficient c can be multiplied with p inside the square root to tune the onset of decay:
image

The steepness can be tuned by using a more generic 1/x^b rather than 1/sqrt(x) (b = 1/2).

image

image

You can play with these parameters [on Desmos](https://www.desmos.com/calculator/mog28vlimx).

### CODE REVIEW

**Comment by kiudee** [comment]
As a reminder: The coefficient is good as it lets us shift the curve on the visit axis. What is still missing is a way to control the steepness.

**Comment by Naphthalin** [comment]
As a comparison, the formula I use in #1173 is
```
1 / (1 + (1/p - 1)*pow(1 + c*n, -a) )
```
which looks a bit more complicated, but basically does the same. This is the formula one gets when treating `p` as a prior in a logit space together with a logarithmic growth in the logit.

Note that `n` can refer both to child and parent visits, `c` should basically to the same as your `c`, and `a` controls the steepness (which should relate to your formula by replacing the `sqrt` with an arbitrary power.

**Comment by Naphthalin** [comment]
Reproducing your 2 graphs from above with the formula from #1173 (`p = [0.02, 0.13, 0.4]`, `c=[0.2, 1, 5]`, `a=0.5`):
april_policies
april_factors

**Comment by AlexisOlson** [comment]
@kiudee That's easy enough to do using 1/x^b instead of 1/sqrt(x).

File: src/mcts/search.cc [hunk 1]
```diff
@@ -301,11 +306,12 @@ std::vector<std::string> Search::GetVerboseStats(Node* node) const {
...
 
-    oss << "(U: " << std::setw(6) << std::setprecision(5) << edge.GetU(U_coeff)
-        << ") ";
+    oss << "(U: " << std::setw(6) << std::setprecision(5)
+        << edge.GetU(U_coeff, childVisits, shift, slope) << ") ";
 
     oss << "(Q+U: " << std::setw(8) << std::setprecision(5)
-        << edge.GetQ(fpu, draw_score, logit_q) + edge.GetU(U_coeff) << ") ";
+        << edge.GetQ(fpu, draw_score, logit_q) +
+           edge.GetU(U_coeff, childVisits, shift slope) << ") ";
```
Comment (Naphthalin) [inline src/mcts/search.cc:314]: `edge.GetU(U_coeff, childVisits, shift, slope) << ") ";`

File: src/mcts/search.cc [hunk 2]
```diff
@@ -253,20 +253,25 @@ std::vector<std::string> Search::GetVerboseStats(Node* node) const {
...
+  const float slope = params_.GetPolicyDecaySlope();
   const bool logit_q = params_.GetLogitQ();
 
   std::vector<EdgeAndNode> edges;
   for (const auto& edge : node->Edges()) edges.push_back(edge);
 
   std::sort(
       edges.begin(), edges.end(),
-      [&fpu, &U_coeff, &logit_q, &draw_score](EdgeAndNode a, EdgeAndNode b) {
+      [&fpu, &U_coeff, &childVisits, &logit_q, &draw_score](EdgeAndNode a, EdgeAndNode b) {
```
Comment (Naphthalin) [inline src/mcts/search.cc:266]: ` [&fpu, &U_coeff, &childVisits, &logit_q, &draw_score, &shift, &slope](EdgeAndNode a, EdgeAndNode b) {`

**Comment by AlexisOlson** [comment]
**To do:** figure out GetVisitsToReachU

**Comment by AlexisOlson** [comment]
Abandoning in favor of:

https://github.com/LeelaChessZero/lc0/pull/1173

https://github.com/LeelaChessZero/lc0/pull/1288

**Pull request closed without merging**